### ISSUE TITLE
[UPDATE] Google Auth Flows and WebView2 

### ISSUE THREAD

**Issue by jasonstephen15** [MicrosoftEdge/WebView2Feedback#1647]
Google has recently made a [policy update](https://developers.googleblog.com/2021/06/upcoming-security-changes-to-googles-oauth-2.0-authorization-endpoint.html?m=1) to prohibit Google OAuth requests in embedded browsers (webviews). This means that google auth flows will not be supported in WebView2.  

Our short-term recommendation for a workaround is to launch the system browser and handle the auth flow there. Google’s OAuth [Sample repo](https://github.com/googlesamples/oauth-apps-for-windows/blob/master/OAuthDesktopApp/OAuthDesktopApp/MainWindow.xaml.cs) has an example of how to do this. This solution will work for all app types.  

Longer term, our suggestion will be to use the [Web Authentication Broker](https://docs.microsoft.com/en-us/windows/uwp/security/web-authentication-broker) (WAB) API. The WAB API is a Windows API, vetted by Google, that will enable auth flows in your native applications. _**This API is currently UWP-only**_ but has plans to be available in win32 and .NET as part of the [WindowsAppSDK](https://github.com/microsoft/WindowsAppSDK). Please follow the team’s [GitHub post](https://github.com/microsoft/WindowsAppSDK/issues/441) for updates on that. For more guidance on how to build out these auth flows visit the [WAB Sample Code](https://github.com/microsoft/Windows-universal-samples/tree/main/Samples/WebAuthenticationBroker). 

Feel free to leave any comments or questions below. Thanks!

**Comment by markismail7** [comment]
This is not a solution. I'm not blaming Microsoft for it. Google is a dictator. We need to move away from Google products. The only way we can do so, if Microsoft created a Gmail. Until then, we are all screwed. The is could get way worse, the sign in with Google is a disaster right now. We can't use it with WebView2.

**Comment by GrandStrateguerre** [comment]
@markismail7 
For me it's already done I don't use Google anymore. I use: https://www.qwant.com/
The only application I use that has an interest for me is Youtube mainly for knowledge sharing. I have forbidden my children to use Google and FaceBook apps.
:)
G.

**Comment by rjx-ray** [comment]
We are CEF users thinking about moving to WebView2, just waiting for https://github.com/MicrosoftEdge/WebView2Feedback/issues/20 to be progressed.

We have faced exactly this authentication issue in CEF and after some considerable effort I managed to get an OAuth access token by spawning an external browser. from our app.
See my post at https://www.magpcss.org/ceforum/viewtopic.php?f=10&t=17457#p45389 with example code attached.

But as it explains in the post I have been unable to find any way of using the token for general web requests.

If anyone here has any ideas that would be very useful

**Comment by markismail7** [comment]
So if I use the OAuth, the user can sign in with the main browser, then once the user is signed in, the user can go back to use Webview2? It's like scratching your ear with your toe SMH. It's like we have nothing to do, but to work extra for google. Imagine every website is doing the same thing. I'm really surprised that most people are ok with it?

**Comment by jarno9981** [comment]
Rrrrrrrrr   is swichtes from cef sharp to webview2 to use google login and now it breaks again i pissed off now

**Comment by jarno9981** [comment]
@markismail7  i hate the security change from google now my app loses lot of funtions

**Comment by federicorosso1993** [comment]
Can you help me with the workaround?
I understand that I need to use Google’s OAuth API on the native browser, and I used that some months ago in a Unity3d game to access google drive and upload/download a save file, but I don't really understand how to do that to be able to be logged in in a browser-like app (my current webview2 don't really just need to access a google service like gmail or drive like in my unity game, I'm using my webview2 as a real browser so I cannot just ask to access gmail o google drive, I need to be able to navigate to any google related service and use it, like youtube).
The example use a Loopback to get the log in response but in the end it just print the response as a string, that's not a solution at all. What we really need (to call this a workaround) is to be able to set that response to the webview2 "cookies/cache/localstorage/whatever google is using to know that the browser is logged in". That's a "good" workaround.
The best workaround would be something even better... a blackbox solution where the webview2 open an edge window->make the user log in inside edge->set whatever google need to know that webview2 is logged into the webview2.

Not sure why you think WAB is a solution here... That's only a solution in the same way Oauth is a solution, when you know how to set the response into webview2 to make it know you are logged in.
Finally I have to say that I find many answer here unrealistic... I don't think any user will leave google just because of this. Still I think their security solutions are just really stupid... there is no reason why I should not be able to log in to my own application in a webview2, if the browser is not supported google should just tell you to log in with an alternative method (like the apps password or something like that). I don't really understand why I cannot even trust my own webview2 browser-like app.
Also when I made that game unity3d and used the google drive api, I was also able to do the same thing with dropbox and microsoft one drive. Dropbox was the best (the docs API is amazing), Microsoft was ok (I needed to contact you on github because the docs were not clear), Google API was the worst and behind everyone (you cannot even make a call without the client-secret even if you use PKCE while everybody else if you use PKCE do not require client-secret and that's a security problem on their Oauth call for desktop app).

**Comment by markismail7** [comment]
@federicorosso1993 I did update to the latest version, still having the same issue. How did you fix it?

**Comment by federicorosso1993** [comment]
> @federicorosso1993 I did update to the latest version, still having the same issue. How did you fix it?

I did not fix it. By solution I mean what should be done by Microsoft, but the problem is still there

**Comment by markismail7** [comment]
Why are we paying the price for Google's problem. When a user attempts to sign in with Google, let the Webview2 opens the Chrome Edge, then after signing in, it'll load automatically back to Webview2. That's a solution. Not make me register my app for something I'm not familiar with. This whole thing is stupid. Real Hackers don't need a browser to attack Google. SMH. I don't think it's about security.

**Comment by haiduong87** [comment]
Here's a practice from gg.
https://github.com/googlesamples/oauth-apps-for-windows

I can't point the different between my own way and this suggestion.
I'm doing as:
- Show a webiview2 web login page
- Handle navigating events to grab the key
- Call rest api to get token

**Comment by federicorosso1993** [comment]
> Here's a practice from gg.
> https://github.com/googlesamples/oauth-apps-for-windows
> 
> I can't point the different between my own way and this suggestion.
> I'm doing as:
> 
> * Show a webiview2 web login page
> * Handle navigating events to grab the key
> * Call rest api to get token

Can you use this sample to navigate on a webview2 after the login like in a regular browser for all the Google website like Gmail and YouTube? 

I would like the user to be able to use my app as a browser and only use the real browser during log in, is that possible? 

I used a little of Google drive API for another project and I know how to ask the user the Auth code from an app and wait for the log in response, but after that I can only make another api call to access Google drive files with my own interface, I cannot set that token in a webview2 to make the user able to navigate in Google drive logged in. 

That's the main problem to solve

**Comment by haiduong87** [comment]
> 
> 
> > Here's a practice from gg.
> > https://github.com/googlesamples/oauth-apps-for-windows
> > I can't point the different between my own way and this suggestion.
> > I'm doing as:
> > 
> > * Show a webiview2 web login page
> > * Handle navigating events to grab the key
> > * Call rest api to get token
> 
> Can you use this sample to navigate on a webview2 after the login like in a regular browser for all the Google website like Gmail and YouTube?
> 
> I would like the user to be able to use my app as a browser and only use the real browser during log in, is that possible?
> 
> I used a little of Google drive API for another project and I know how to ask the user the Auth code from an app and wait for the log in response, but after that I can only make another api call to access Google drive files with my own interface, I cannot set that token in a webview2 to make the user able to navigate in Google drive logged in.
> 
> That's the main problem to solve

That's not my use case, so I haven't tested that.

But I think it's ok,

About the HttpListener, I have never used that before, may be there's firewall problem.

**Comment by federicorosso1993** [comment]
> > > 
> 
> That's not my use case, so I haven't tested that.
> 
> But I think it's ok,
> 
> About the HttpListener, I have never used that before, may be there's firewall problem.

Don't really know what are you talking about for the httplistener, I don't have a problem with that in my other project. The problem is only with my last project where I use a webview2 as a browser and there is no way to log in to Google and watch YouTube videos after that in my own browser like app. If I can only make a oauth2 call to get the video Flux and watch it in my own interface that will only work for YouTube but not for Google drive or Gmail then there is no reason to use a webview in the first place. Of course I cannot only use my interface because I also need my browser to work on all website (not just Google services). It's crazy how difficult they made it just because they refuse to make an alternative for the Google log in... It's not even possible to use a password application like you can do for older outlook.

**Comment by Beej126** [comment]
> > Here's a practice from gg.
> > https://github.com/googlesamples/oauth-apps-for-windows
> > I can't point the different between my own way and this suggestion.
> > I'm doing as:
> > 
> > * Show a webiview2 web login page
> > * Handle navigating events to grab the key
> > * Call rest api to get token
> 
> Can you use this sample to navigate on a webview2 after the login like in a regular browser for all the Google website like Gmail and YouTube?

@federicorosso1993, yes [this sample](https://github.com/googlesamples/oauth-apps-for-windows/tree/master/OAuthDesktopApp/OAuthDesktopApp) works fine with webview2 and doesn't require any additional google login for gmail, youtube, etc... it's easy to confirm yourself as i have... add webview2 control (with Name="wv2") to bottom of existing window, then replace [the one line ](https://github.com/googlesamples/oauth-apps-for-windows/blob/3f14353db7085ea1927ae6bc9951f39bf6946ec7/OAuthDesktopApp/OAuthDesktopApp/MainWindow.xaml.cs#L84) that spawns external browser:
```
System.Diagnostics.Process.Start(authorizationRequest);
```
with webview2 equivalent:
```
wv2.Source = new Uri(authorizationRequest);
```

it appears the webview2 browser instance maintains full fidelity... we can hit F12 to open typical chrome debugger tools window and inspect that google's login cookies are indeed present.

@jasonstephen15 , @markismail7, @jarno9981, @64Soft, @Simon4638 - i'm mainly posting because i want to better understand what roadblocks you are running into, so i don't waste time going down a dead end for my own project...

it seems embedded browser approach is still functionally viable... so is this entirely about the change in google's policy? i.e. we're running afoul of rules if we continue to use embedded, even if it works?

maybe latest webview2 (v1.0.992.28 stable, published 2021-09-27) closed previous gaps?

**Comment by federicorosso1993** [comment]
@Beej126 thank you for trying it. I'll try to undestand how to implement it in my own app.

**Comment by jarno9981** [comment]
@Beej126 thank i am going to try to

**Comment by jarno9981** [comment]
@Beej126 still doesn't log me in only on edge browser

.
{
  "sub": "100246938796268582855",
  "name": "Jarno",
  "given_name": "Jarno",
  "picture": "https://lh3.googleusercontent.com/a-/AOh14Ghy0fzITDxtGU_GT6Qw75eTyPTDuK3yaqDXqtReiA\u003ds96-c",
  "locale": "nl"
}


and returns this in the textbox of the browser

but when i try to login it says unsafe app  Winforms

**Comment by jarno9981** [comment]
after email in external browser
![Schermafbeelding 2021-10-05 070948](https://user-images.githubusercontent.com/53493418/135963812-95854cbf-70cc-42c4-bf71-f67d04c6ae9f.png)
after trying to login in email
![Schermafbeelding 2021-10-05 071019](https://user-images.githubusercontent.com/53493418/135963814-f8849c7c-5fa3-4af7-814d-a6d228fa8f7c.png)

**Comment by Beej126** [comment]
if i understand you, you're hoping that embedded webview2 login is shared with external browser... that might be possible with further effort but that's not going to work by default ... i believe they are running under separate "user profiles", i.e. a different bucket of all the browser state, cookies, etc

i misunderstood your desired scenario... what i am looking for, and what i thought you all were looking for, was that after you use webview for google oauth flow, you can also then count on the google auth login to be present under the webview for other google sites, youtube, gmail, etc

**Comment by jarno9981** [comment]
@Beej126 I want to by abble to login from winforms app that correct

**Comment by jarno9981** [comment]
when redirect url is active this shows
![Schermafbeelding 2021-10-05 072109](https://user-images.githubusercontent.com/53493418/135964827-e2a7f976-5fe5-4ca5-be47-e1cee56cd2f6.png)

**Comment by jarno9981** [comment]
this after email clicked
![Schermafbeelding 2021-10-05 072233](https://user-images.githubusercontent.com/53493418/135964966-e36deeb3-26b0-4a30-96ec-3a5c634d5a97.png)

**Comment by Beej126** [comment]
i guess you're not doing what i said, to modify the sample to use the webview2?
i'll post a fork of the project with what i mean here in a few mins.

**Comment by Beej126** [comment]
this fork has the embedded webview tweak i'm suggesting in the wpf sample: https://github.com/Beej126/oauth-apps-for-windows

first image shows the oauth flow in the embedded webview
second image shows i can also jump right into youtube after the oauth flow (without logging in again)
![image](https://user-images.githubusercontent.com/6301228/135968815-c85a3ddc-e2bd-4e3b-8477-cda7cc058302.png)
![image](https://user-images.githubusercontent.com/6301228/135968992-d114cc45-8537-4a98-8fd2-aa2104183cf9.png)

**Comment by jarno9981** [comment]
@Beej126 i will try that like that your doing

**Comment by jarno9981** [comment]
returns this for me

![Schermafbeelding 2021-10-05 092052](https://user-images.githubusercontent.com/53493418/135977928-2c9ea515-48e5-42ca-adf7-c55e5bd09a61.png)

no refresh

**Comment by Beej126** [comment]
try running my sample as-is and let us know what happens.

**Comment by jarno9981** [comment]
![Schermafbeelding 2021-10-05 093848](https://user-images.githubusercontent.com/53493418/135980548-2058d6a7-79b5-4522-b00d-3078648b20c9.png)

sample app returns same for me

**Comment by Beej126** [comment]
ok now that's interesting!
i am trying to think of variables in play here and i remember seeing something about the running webview2 binary is running on top of the version of msedge you have installed... i'm running msedge 64bit v94.0.992.38... also make sure to at least try loading the msedge DEV build, i remember seeing something about that dependency.

since you're running my project, you should have 1.0.1018-prerelease of webview2 assembly but double check that.
here's the release notes page that ties versions of webview2 to msedge.
https://docs.microsoft.com/en-us/microsoft-edge/webview2/release-notes
> For full API compatibility, this prerelease version of the WebView2 SDK requires Microsoft Edge version 95.0.1018.0 or higher... 

which is interesting, i just updated msedge and didn't get v95, so that could be a DEV build#, which means you should try installing latest DEV build msedge

the only other thing i can think of right now is i am running Win11 (version 21H2, build 22000.194)... but i don't think that's a dependency here

thanks for sharing that screenshot, it really proves (for everybody) there is something odd going on here.

you should also try fiddling with:
```
wv2.CoreWebView2.Settings.UserAgent = "copy this from working browser";
```
https://stackoverflow.com/questions/67549605/set-useragent-in-webview2/67550089#67550089

watch for differences like that comparing the debug windows of a working browser and not working webview2 (remember f12 opens debug)... i feel like webview2 might be sending a useragent that's tripping off google, but just a wild hunch ... since i can't repro on my end i can't help you here.

better yet, just trying temporarily spoofing the useragent string in debug tools and hit refresh:
https://developer.chrome.com/docs/devtools/device-mode/override-user-agent/

here's a page that shows what user agent is being sent, plug this into webview2 to see what it's sending by default:
https://www.whatismybrowser.com/detect/what-is-my-user-agent

let me know how that goes

here's mine:

![image](https://user-images.githubusercontent.com/6301228/136045570-e34e227d-cba8-4d2f-a8eb-741a1ee5c86e.png)

**Comment by jarno9981** [comment]
@Beej126  my useragent is

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.71 Safari/537.36 Edg/94.0.992.38

**Comment by Beej126** [comment]
ok well scratch that idea, that looks totally normal... i would definitely upgrade all the msedge stack including installing msedge dev build to cross that angle off your list.

total longshot but if you're not on win11 and desperate i would try spinning up a win11 vm and running my project to see if any difference.

**Comment by jarno9981** [comment]
@Beej126 
I am on win11 bete channel

I am going to try edge dev next

**Comment by jarno9981** [comment]
@Beej126  even with edge dev same result unable to login not safe error 

Winforms

**Comment by Beej126** [comment]
- reading up on the MiM attack potential helps me appreciate why google is doing this, so i'm going to migrate away from the doing the auth in embedded... 
- but i'm still wondering what i should expect beyond the context of oauth for making api calls ... 
- i wish i was getting the same behavior as you so i knew what was and wasn't possible... 
- is your webview2 blocked from directly hitting all google pages that require auth, ie. gmail, youtube, etc?
* this is a good discussion: https://stackoverflow.com/questions/41801836/how-can-google-detect-a-request-is-from-a-webview
* besides user agent, they mention a header called "X-Requested-With"...
* it would be interesting to see your request headers, even your whole debug network tab for that matter... mine attached for example - **DONT FORGET TO REDACT YOUR COOKIES AND ANYTHING THAT LOOKS SENSITIVE**

![image](https://user-images.githubusercontent.com/6301228/136106926-f35211e8-6187-49ac-bd68-cdb2e5eeec15.png)

general
----------
Request URL: https://accounts.google.com/o/oauth2/v2/auth?response_type=code&scope=openid%20profile&redirect_uri=http%3A%2F%2F127.0.0.1%3A9810%2F&client_id=581786658708-elflankerquo1a6vsckabbhn25hclla0.apps.googleusercontent.com&state=83_ClWKyPX8s__FAFSjIC5noEie8YbZC7c--4umkZLc&code_challenge=Cn470IYGFw-bO11du2LJtPYoCYonZVLLNMWofsjhiT4&code_challenge_method=S256
Request Method: GET
Status Code: 200 
Remote Address: 172.217.14.205:443
Referrer Policy: strict-origin-when-cross-origin

response headers
----------
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
cache-control: no-cache, no-store, max-age=0, must-revalidate
content-encoding: gzip
content-security-policy: script-src 'report-sample' 'nonce-AG3on8/fwioIY4p51neqyg' 'unsafe-inline' 'unsafe-eval';object-src 'none';base-uri 'self';report-uri /cspreport
content-type: text/html; charset=utf-8
date: Tue, 05 Oct 2021 19:00:08 GMT
expires: Mon, 01 Jan 1990 00:00:00 GMT
pragma: no-cache
server: GSE
set-cookie: [redacted]
strict-transport-security: max-age=31536000; includeSubDomains
x-content-type-options: nosniff
x-frame-options: DENY
x-xss-protection: 1; mode=block

request headers
----------
:authority: accounts.google.com
:method: GET
:path: /o/oauth2/v2/auth?response_type=code&scope=openid%20profile&redirect_uri=http%3A%2F%2F127.0.0.1%3A9810%2F&client_id=581786658708-elflankerquo1a6vsckabbhn25hclla0.apps.googleusercontent.com&state=83_ClWKyPX8s__FAFSjIC5noEie8YbZC7c--4umkZLc&code_challenge=Cn470IYGFw-bO11du2LJtPYoCYonZVLLNMWofsjhiT4&code_challenge_method=S256
:scheme: https
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9
cookie: [redacted]
sec-ch-ua: "Chromium";v="94", "Microsoft Edge";v="94", ";Not A Brand";v="99"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
sec-fetch-dest: document
sec-fetch-mode: navigate
sec-fetch-site: none
sec-fetch-user: ?1
upgrade-insecure-requests: 1
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.71 Safari/537.36 Edg/94.0.992.38

query string parms
----------
response_type: code
scope: openid profile
redirect_uri: http://127.0.0.1:9810/
client_id: 581786658708-elflankerquo1a6vsckabbhn25hclla0.apps.googleusercontent.com
state: [redacted]
code_challenge: [redacted]
code_challenge_method: S256

**Comment by jarno9981** [comment]
@Beej126 

for me it returns this
![Schermafbeelding 2021-10-06 110119](https://user-images.githubusercontent.com/53493418/136173443-13b10a73-9637-47ae-afc7-13057acd20a3.png)

**Comment by jarno9981** [comment]
this in headers 
![Schermafbeelding 2021-10-06 111154](https://user-images.githubusercontent.com/53493418/136174419-d919f172-81fd-4215-9563-6c4fb4dae2a7.png)

**Comment by Beej126** [comment]
it would be the request headers going off to google that they could pick up on and decide block you (i only see response headers in that screenshot)

but help me know something for my own needs - can you load youtube into my sample app and login via the normal flow present on the page itself?

(use the app menu icon in the upper right corner of google.com to select youtube.)
![image](https://user-images.githubusercontent.com/6301228/136214637-fadbf7f0-aeab-4f55-b1fc-1027c41238bd.png)

**Comment by federicorosso1993** [comment]
@Beej126 I tried your code with my own client id, secret id (btw you should not share those, but you need to tell what scopes did you use).
It's not working and I don't really see how can it work for you.
On the first solution (System.Diagnostics.Process.Start(authorizationRequest);) you will only get a text response (that works of course) on an external browser. So the OAuth2 part it's working like it should.
But once you go back to the app the webview2 it's not logged (of course).
On the second solution it redirect you correctly to the API app but you will not be able to log in like in the regular log in page....
There is no difference between trying to log in on the regular google page 
https://accounts.google.com/signin/v2/identifier?flowName=GlifWebSignIn&flowEntry=ServiceLogin
and the one that you use when you do that with the oauth2 code
https://accounts.google.com/o/oauth2/v2/auth/oauthchooseaccount?response_type=code&etc....
They both say your browser is not supported for security reason after you input your email and you try to press the Continue button
so you can get here:
https://user-images.githubusercontent.com/6301228/135968815-c85a3ddc-e2bd-4e3b-8477-cda7cc058302.png

But you will not get after that.

Not sure why you can... maybe you only got a cache on your webview2 or maybe the windows 11 preview was still not blocked when you posted the message that's the only way I can explain this.

What we need to understand is once you make the API call to an external browser (like you did with the first solution) and you get a response (and we need to understand what scopes are needed there for all google services) how to set that info into the cookies / localstorage or something like that to be able to sign in on the webview2.
That's the main point to solve, your first solution is not complete and the second solution does not work.

**Comment by federicorosso1993** [comment]
@Beej126 Already checked, it's the same.

**Comment by Beej126** [comment]
@markismail7 - thanks for chiming in...  i saw you on the other thread and roped you in here because i'm wanting to hear from as many different experiences as possible, but the external browser token auth flow DOES work fine and isn't the problem here... 

what i'm looking for (and i believe others on this thread), is having an **embedded browser in the native app that lets you get into google pages that require auth**, gmail, youtube, etc... i am hearing this is currently being blocked by google for everybody??... except me somehow, for the moment... i'm keen to understand why so that i can count on this and not build stuff that then breaks when mine stops working as well.

**Comment by jarno9981** [comment]
@Beej126 which webview2 runtime pakage do you have maybe thats de problem for me

**Comment by Beej126** [comment]
@jarno9981 - i'm seeing the same working behavior in google's wpf sample with webview2 v1.0.992.28 (which is a .net framework 4.6.1 project)  as well as another fresh .net 5 pet project with webview2 v1.0.1018-prerelease. 

i have currently msedge v94.0.992.38 and msedge dev v96.0.1032.0... the [release notes](https://docs.microsoft.com/en-us/microsoft-edge/webview2/release-notes) talks about webview2 browser version dependencies.

@federicorosso1993 - thanks for setting me straight on your mention of what step in the oauth flow, that makes sense now.

i'm curious how much anyone has explored the cookie manager api?
https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2experimentalcookiemanager?view=webview2-1.0.674-prerelease#getcookies

**Comment by gplwzz1989** [comment]
You are all geniuses，Settings.UserAgent = "Chrome" Settings.UserAgent = "Andriod" ，It's working

**Comment by jarno9981** [comment]
@Beej126  @gplwzz1989    when you login you get this after google url now i think they are working on it 

google.com/spf#=121676

got that yesterday

**Comment by jarno9981** [comment]
![Schermafbeelding 2021-11-23 101942](https://user-images.githubusercontent.com/53493418/142998596-1ff85b3e-4b34-44b6-ade2-bc27a8595530.png)

this result after logging in

**Comment by ono77** [comment]
Hi, today If I access to Youtube I see error This Page isn't working, www.youtube.com redirected you too many times. Try cleaning your cookies.
If I access with Edge I see the cookie notice window that has to be accepted...

**Comment by jarno9981** [comment]
> Hi, today If I access to Youtube I see error This Page isn't working, [www.youtube.com](http://www.youtube.com) redirected you too many times. Try cleaning your cookies. If I access with Edge I see the cookie notice window that has to be accepted...

updating webview2 runtime will work i had the same problem just re-install or update

**Comment by nirbil** [comment]
Google is a bully and is using their monopolistic position both in browser share and internet traffic to squeeze competition out of the market, and hurt Microsoft whenever they can. Anything's legit to preserve browser market share, as a proxy for maintaining search hegemony.
Notice how they are attempting to block only Windows based embedded Webviews. Android based (which enable the same types of tweaks) are left alone.

**Comment by jarno9981** [comment]
@everyone i found a work around way to use your costum api , thanks to @Beej126  for the basisc code 

after reworking it a little now i can use google api as my login and redirect to the correct page

but the useragent things is still required but now you can use your own api key

**Comment by jarno9981** [comment]
> this fork has the embedded webview tweak i'm suggesting in the wpf sample: https://github.com/Beej126/oauth-apps-for-windows


reply from this code

**Comment by BobbyShoe** [comment]
It appears like the provided work around no longer works.  I tried also modifying the sample to set the user-agent to Chrome, but that did not help either.

Anyone have any other suggestions?  Has anyone had any luck trying to possibly store the returned Google OAuth token somewhere in the WebView2 profile cache, so the return OAuth token could be used for requests to other sites?

![image](https://user-images.githubusercontent.com/7215411/157591831-a6ca9eca-1f66-498b-a4ef-56c0b2b65f73.png)

**Comment by haiduong87** [comment]
> It appears like the provided work around no longer works. I tried also modifying the sample to set the user-agent to Chrome, but that did not help either.
> 
> Anyone have any other suggestions? Has anyone had any luck trying to possibly store the returned Google OAuth token somewhere in the WebView2 profile cache, so the return OAuth token could be used for requests to other sites?
> 
> ![image](https://user-images.githubusercontent.com/7215411/157591831-a6ca9eca-1f66-498b-a4ef-56c0b2b65f73.png)

Please check this:  
https://github.com/MicrosoftEdge/WebView2Feedback/issues/1647#issuecomment-952826012

And if you want to test with above sample
- Press "Try Again"
- F12 -> modify UserAgent (Network Conditions) -> Keep the dev tool open and test again.

**Comment by jarno9981** [comment]
> It appears like the provided work around no longer works. I tried also modifying the sample to set the user-agent to Chrome, but that did not help either.
> 
> Anyone have any other suggestions? Has anyone had any luck trying to possibly store the returned Google OAuth token somewhere in the WebView2 profile cache, so the return OAuth token could be used for requests to other sites?
> 
> ![image](https://user-images.githubusercontent.com/7215411/157591831-a6ca9eca-1f66-498b-a4ef-56c0b2b65f73.png)

Do You set the useragent on navigating?

**Comment by BobbyShoe** [comment]
@jarno9981 thanks for the suggestion.  Setting it on NavigationStarting did work.  I was previously setting the user agent, before setting the Source, but that did not appear to work for some reason.

Does anyone now if there is anyway to use the returned token obtained outside the WebView2 as well (from the original unmodified demo)?  This will functionally work, but appears to be in conflict with the Google documented usage (because it is still running in a WebView to authenticate).  I am concerned some future Google change could break this.

Although not an ideal solution it seems like the following approach would be compliant with the Google documented usage for the OAuth API

1) Have a WebView2 open (that the user is navigating)
2) When Google Access is requested, launch a system installed browser to capture authentication information (this seems to be the "compliant" approach).
3) Get the returned token from the Google OAuth callback.
4) Inject the token into the WebView2 instance for use on sites requiring Google Auth.

Thanks again for the feedback.

**Comment by BobbyShoe** [comment]
I have been trying to look into what a typical Google Sign in does after authenticating.  It actually looks like Google is automatically authenticating against the Google.com and YouTube.com domains after a login.

After authenticating a GET request is sent to 
https://mail.google.com/accounts/SetOSID  (I was trying to login to GMail for the test, hence mail.)

and a request is sent to
https://accounts.youtube.com/accounts/SetSID

I think these they calls are then setting all of the authentication cookies that are needed for all Google Authentication.  For example after I logged in, I could hit Google Drive without having to login again, it looks like the authentication cookies are being sent over with the request.

Has anyone tried using the returned tokens from the Google OAuth callback to call these two URLs?  I am not quite sure what all of the parameters are supposed to be.

It looks like calling these two URLs within a WebView2 instance with a token received from an external system browser login to Google would be in compliance with the Googles terms of use for OAuth.

**Comment by vladimir-ikryanov** [comment]
JFYI: if you configure WebView2 to use the remote debugging port switch, then Google Sign won't work.

**Comment by tobq** [comment]
@vladimir-ikryanov @jarno9981 is it now possible to sign in to google using an embedded browser?

**Comment by Beej126** [comment]
> @vladimir-ikryanov @jarno9981 is it now possible to sign in to google using an embedded (WebView2) browser?

yes i've been doing this for months now with solid results.
my code is here => https://github.com/Beej126/GenericWebAppWpfWrapper
pay particular attention to these methods and comments: [one](https://github.com/Beej126/GenericWebAppWpfWrapper/blob/f2ec4ffdc037d63fd170ac9f5bfdf692c23be555/GenericWebAppWpfWrapper/MainWindow.xaml.cs#L98), [two](https://github.com/Beej126/GenericWebAppWpfWrapper/blob/f2ec4ffdc037d63fd170ac9f5bfdf692c23be555/GenericWebAppWpfWrapper/MainWindow.xaml.cs#L108)

**Comment by tobq** [comment]
> > @vladimir-ikryanov @jarno9981 is it now possible to sign in to google using an embedded (WebView2) browser?
> 
> yes i've been doing this for months now with solid results. my code is here => https://github.com/Beej126/GenericWebAppWpfWrapper pay particular attention to these methods and comments: [one](https://github.com/Beej126/GenericWebAppWpfWrapper/blob/f2ec4ffdc037d63fd170ac9f5bfdf692c23be555/GenericWebAppWpfWrapper/MainWindow.xaml.cs#L98), [two](https://github.com/Beej126/GenericWebAppWpfWrapper/blob/f2ec4ffdc037d63fd170ac9f5bfdf692c23be555/GenericWebAppWpfWrapper/MainWindow.xaml.cs#L108)

It seems the main change required is changing the user-agent?

**Comment by Beej126** [comment]
yes it's been about user-agent since the beginning... but there's a little orchestration required to have it set to something legit like chrome most of the time (see comments for references to legit user-agent strings) ... but then you just need to flip it to something NOT legit when the url is accounts.google.com, i.e. the actual login page... i guess you're saying you can't read C#? i encourage you to just try reading it like english, it's pretty declarative, there is a little nuance to the things i found necessary... besides the user agent, the other thing is trapping the new window event and sending that url back into the existing window... odds are probably low there's any other working examples not in C#... best of luck.

**Comment by hershkoy** [comment]
For those who managed to login, are you experiencing random log out from the gmail session?
We managed to do login with the user agent solution, but after some time, the gmail session is logged out. I really would appreciate to know if there is a solution for this (or to know if I am the only one experiencing it)

**Comment by mikeduglas** [comment]
I haven't experienced logging out of my Gmail session.

--
Mike


чт, 18 авг. 2022 г. в 21:11, hershkoy ***@***.***>:

**Comment by Beej126** [comment]
it stays logged in for multiple days for me... every once in a while i'll notice i have to log back in but it's very infrequent

**Comment by vladimir-ikryanov** [comment]
@Beej126 same in my case. I keep my Gmail tab open 24/7 and from time to time (once a week) Google asks me to log in again.

**Comment by JohnDizzle** [comment]
From deployment I use a UserAgent = "Chrome", and save it to my Settings, and then login to google.   I change UserAgent (To Chrome's Current UserAgen) in my Settings and then restart app and all is fine.

**Comment by Vladoses** [comment]
hello, still no normal solution to this problem? if so can you tell me where

**Comment by ivfit** [comment]
What happens when you remove "Edg/..." part from the WebView2 user agent?

**Comment by jarno9981** [comment]
> What happens when you remove "Edg/..." part from the WebView2 user agent?

No idea try it

**Comment by champnic** [comment]
@jarno9981 That `coolblue.nl` website worked fine for me from a couple of WebView2 apps.

**Comment by jarno9981** [comment]
> @jarno9981 That `coolblue.nl` website worked fine for me from a couple of WebView2 apps.

For me it gets blocked and on edge en chrome it works fine

**Comment by t1soft** [comment]
google login works for me, maybe webview2 updated?

**Comment by jarno9981** [comment]
> google login works for me, maybe webview2 updated?

Does it work without any workaround

**Comment by jarno9981** [comment]
> @jarno9981 That `coolblue.nl` website worked fine for me from a couple of WebView2 apps.

@champnic was my mistake the useragent was messed up

**Comment by t1soft** [comment]
> > google login works for me, maybe webview2 updated?
> 
> Does it work without any workaround @t1soft

@jarno9981 From my experience, it does work, however I don't know if others can log in without workaround

On two computers, I was able to log into google through webview2 without using any workaround. Gmail works alright, have yet to experience random log-outs.
